### ISSUE TITLE
.git-ftp-include files are not being pushed

### ISSUE THREAD

**Issue by CLL80** [git-ftp/git-ftp#283]
Hello again,

I have 2 different syncroots using scope, and a couple files in the `.git-ftp-include`. I'm using version 1.3.0-UNRELEASED. These files are not being pushed no matter what I do.

My config (minus passwords and users):
```
[git-ftp]
	url = ftp://ftp.domain.com
	syncroot = html/
[git-ftp "dev"]
	url = ftp://photos.domain.com/dev
[git-ftp "master"]
	url = ftp://photos.domain.com
[git-ftp "inc"]
	syncroot = include/
[git-ftp "devinc"]
	syncroot = include/
```

This is what I have in my `.git-ftp-include`:
```
!_inc/config.inc.php
vendor/:composer.lock
```
But I have tried many different paths in this file, like:
```
!include/_inc/config.inc.php
include/vendor/:composer.lock
```
and
```
!include/_inc/config.inc.php
!include/vendor/
```
At this point, with the `.git-ftp-include` containing the above, the files in `include` were attempted to be pushed to `html` when pushing the `dev` scope, which, when looking at my syncroots, doesn't make sense to me and is not what I want.

From what I read in the manual, 

> When using the `--syncroot` option, all paths are relative to the set syncroot.

This is ideal for me if it works because I don't want the files in `include` to be pushed to `html`.
Nothing else I do will get these files to be uploaded when I do `git-ftp push -s inc` or `git-ftp push -s devinc` I have changed the composer.lock file. I have removed the `git-ftp.log` from the server and started over with `init`, etc.

My directory structure is like this:
```
project--- html------ files
     | ---include----- vendor---- files
             |-------- _inc ----- files
```

 Am I doing something wrong?

**Comment by plashenkov** [comment]
I’m using git-ftp 1.2.0 on Ubuntu and I’m trying too to use .git-ftp-include with --syncroot option. My syncroot is `public_html`. Everything inside it I need to upload to url: `ftp://x.x.x.x/public_html`, plus folder `public_html/app/vendor`, which is gitignored, when `composer.lock` is changed. `composer.lock` is located in the root of git repository.

I.e. my folder structure:
```
.git-ftp-include
composer.json
composer.lock
public_html
  |__app
       |__vendor
```

[Git-ftp documentation](https://github.com/git-ftp/git-ftp/blob/develop/man/git-ftp.1.md#syncing-untracked-files) says that folders inside `.git-ftp-include` become relative to syncroot (which is strange because `.git-ftp-include` file is located in root, but folders listed inside it must be specified relatively to syncroot).

Here is the content of `.git-ftp-include`:

```
app/vendor/:/composer.lock
```

And this does not work. Everything else is uploaded very nicely, but not the vendor folder. I’ve also tried to write paths in different formats:

```
public_html/app/vendor/:composer.lock
/public_html/app/vendor/:/composer.lock
/public_html/app/vendor:/composer.lock
etc
```
as well as move `.git-ftp-include` to `public_html` (which is a syncroot) folder.

Nothing of this works.
What do I do wrong?

Thank you in advance.

**Comment by mkllnk** [comment]
I agree that it is confusing. Some of that behaviour has also changed between versions. The thing with prefixing a `/` for files outside the syncroot is actually not mentioned in the manual of version 1.2.0 and it was not implemented back then. You are doing nothing wrong, it is just not supported.

You can try the newest development version, but as you see above, it could have some bugs.

**Comment by plashenkov** [comment]
Hi Maikel! Thank you for your answer.

As I think it’s better to have paths listed inside `.git-ftp-include` to be relative to that file and not to have confusing rules with syncroot and additional slashes. A user can use several different syncroots in one project, in this case `.git-ftp-include`—which must reside in git repository root, if I am not mistaken—will work for one syncroot and will not for other. We should use `.git-ftp-include` relatively to repository because it resides in repository root and it is expected to be git repository global.

As for the manual, it is naturally expected as well that the default manual here, on project’s Github page, matches the last stable version which is found on the [releases page](https://github.com/git-ftp/git-ftp/releases). Currently, to get actual manual for the stable version, I must expand branches/tags and specially select 1.2.0 branch. The [site also points me](https://git-ftp.github.io/#further-reading) to the dev version manual. Maybe it is better to have the default branch pointed to the latest stable. Sorry if I missed something.

**Comment by mkllnk** [comment]
Thank you for the feedback. That sounds very reasonable. We should probably introduce a master branch with the latest stable version again.

**Comment by mkllnk** [comment]
@CLL80 Thank you for bringing this up. I just tested this and the manual seems to be wrong. Including works with the syncroot in the include file. Passing test setup:

```sh
mkdir "public_html"
echo "public_html" > .gitignore
echo "content" > "public_html/always.html"
echo "!public_html/always.html" > .git-ftp-include
git add .
git commit -m "setup"
git-ftp push --syncroot "public_html"
# always.html is uploaded
```

**Comment by CLL80** [comment]
Thanks for taking the time to look into this and for confirming. However, I am seeing some strange behavior that, to me, doesn't seem to be just the manual.

I can confirm that when putting the full path in the `.git-ftp-ignore`, it does get pushed, but it also causes that directory to be uploaded when using the other syncroot.

I would think anything inside `include` would be ignored when pushing using the `html` syncroot (they are siblings), but these files are being pushed even when specifying the syncroot that doesn't contain them.

For example, when doing `git-ftp push -s inc --syncroot "include"`, all the files that are in the `.git-ftp-include` file  inside the `include` directory are being pushed, as expected. This is the relevant part of the config for that scope:

```
[git-ftp "inc"]
	syncroot = include/
```

But when doing `git-ftp push -s dev --syncroot "html"`, all the same files that are inside `include` (outside the `html` syncroot I am pushing) are being pushed still. Here is the relevant part of the config for that scope:

```
[git-ftp "dev"]
	syncroot = html/
```

**Comment by mkllnk** [comment]
> I can confirm that when putting the full path in the `.git-ftp-ignore`, it does get pushed, but it also causes that directory to be uploaded when using the other syncroot.

Did you mean `.git-ftp-include`?

If you include based on a file outside the syncroot like `/composer.lock` then the include algorithm will ignore if the target file is in the syncroot or not. And some people probably want that behaviour.

I tested only half of the problem before. The target file name needs the syncroot, but the source file does not. Here is another working example:
```bash
mkdir "public_html"
echo "content" > "public_html/style.sass"
echo "content" > "public_html/style.css"
echo "public_html/style.css" > .gitignore
echo "public_html/style.css:style.sass" > .git-ftp-include
git add .
git commit -m "setup"
git-ftp push --syncroot "public_html"
# style.css is uploaded
```

Maybe it would be good to support arbitrary hooks that can add or remove files from the list before it's pushed. Not sure if I will have time for a bigger change of the code. Pull requests are always welcome.

**Comment by CLL80** [comment]
Yes, I did mean `.git-ftp-include`.

At this point, I'm not including based on any files—just static includes using `!`. When I have includes like this:
```
!include/_inc/config.inc.php
!include/vendor/
```

They are pushed using the `include` syncroot as intended, like this:
`git-ftp push --syncroot include/ -s devinc`

But my problem is, they are also uploaded when pushing with other scopes with other destination directories, like `html`, in my case. This happens even without basing them on a file outside the syncroot like `composer.lock`. I don't think this is intended unless I am misunderstanding your explanation.

Is there currently a way to have files/directories in the `.git-ftp-include` that will only be pushed for one scope? I was thinking I could accomplish this by having them in a separate directory that resides at the same level as the other scope's syncroot.

**Comment by mkllnk** [comment]
> Is there currently a way to have files/directories in the .git-ftp-include that will only be pushed for one scope?

No, not really. You could have different branches with different .git-ftp-include files. Or you define your own commands that link .git-ftp-include to the rules you want and then do the push.

**Comment by explorier** [comment]
I couldn't quite figure this out - on git-ftp version 1.0.2 at first, then updated to git-ftp version 1.3.3 and tried it, neither worked for me. All other files are uploading fine.

By CLL80's last comment I understand that the git-ignored files pushed successfully when explicitly including the syncroot in the .git-ftp-include file declaration, but this isn't working for me. I tried writing the path from the git repo root, as well as one level up from the syncroot (taking out "themes"). Any ideas?

Here's the relevant part of my .git/config:

`
[git-ftp "dha-theme"]
url = ftp://ftp.mysite.com:21/my.site/wp-content/themes/
user = ***@*****.com
password = *****
syncroot = site/web/app/themes/
`
and my .git-ftp-include
```
!themes/alchemy-v2/dist/

```

**Comment by mkllnk** [comment]
@explorier The path from the git repository root should be correct:

    !site/web/app/themes/alchemy-v2/dist/
